pkg/mcclient/listwatch/listwatch.go

```diff
@@ -0,0 +1,108 @@
+package listwatch
+
+import (
+	"context"
+	"fmt"
+
+	machmeta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	clientscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	upcache "k8s.io/client-go/tools/cache"
+	log "k8s.io/klog/v2"
+)
+
+type CrossClusterListerWatcher struct {
+	clusterListWatch map[string]*upcache.ListWatch
+}
+
+func NewCrossClusterListerWatcher(clusterListWatch map[string]*upcache.ListWatch) *CrossClusterListerWatcher {
+	clw := &CrossClusterListerWatcher{
+		clusterListWatch: clusterListWatch,
+	}
+	return clw
+}
+
+func (clw *CrossClusterListerWatcher) Watch(options metav1.ListOptions) (watch.Interface, error) {
+	combinedCh := make(chan watch.Event)
+	for clusterName, lwForCluster := range clw.clusterListWatch {
+		log.Infof("--- Whatch for cluster: %s", clusterName)
+		clusterWatch, err := lwForCluster.Watch(options)
+		if err != nil {
+			return nil, fmt.Errorf("Watch for cluster %s failed: %w", clusterName, err)
+		}
+
+		go func(clusterName string) {
+			for {
+				select {
+				case <-context.Background().Done(): //TODO: pass the ctx as param?
```

Comment:
(1) https://github.com/kubernetes/client-go/blob/v0.27.2/tools/cache/listwatch.go#L39 takes no context, so I think you want to follow suit.

(2) `context.Background()` is never done.

(3) You might want an early-out for other reasons, such as https://github.com/kcp-dev/edge-mc/pull/561/files#r1220068089